File: app/src/main/java/com/tematikhonov/androidgeekbrainssecondapp/ui/NoteFragment.java
package com.tematikhonov.androidgeekbrainssecondapp.ui;

import android.content.Context;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.DatePicker;

import com.google.android.material.textfield.TextInputEditText;
import com.tematikhonov.androidgeekbrainssecondapp.MainActivity;
import com.tematikhonov.androidgeekbrainssecondapp.R;
import com.tematikhonov.androidgeekbrainssecondapp.data.Note;
import com.tematikhonov.androidgeekbrainssecondapp.data.Publisher;

import java.util.Calendar;
import java.util.Date;

public class NoteFragment extends Fragment {

    private static final String ARG_CARD_DATA = "Param_CardData";

    private Note note;
    private Publisher publisher;
    private TextInputEditText title;
    private TextInputEditText description;
    private DatePicker datePicker;

    public static NoteFragment newInstance(Note note) {
        NoteFragment fragment = new NoteFragment();
        Bundle args = new Bundle();
        args.putParcelable(ARG_CARD_DATA, note);
        fragment.setArguments(args);
        return fragment;
    }

    public static NoteFragment newInstance() {
        NoteFragment fragment = new NoteFragment();
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            note = getArguments().getParcelable(ARG_CARD_DATA);
        }
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        MainActivity activity = (MainActivity)context;
        publisher = activity.getPublisher();
    }

    @Override
    public void onDetach() {
        publisher = null;
        super.onDetach();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_note, container, false);
        initView(view);
        if (note != null) {
            populateView();
        }
        return view;
    }

    @Override
    public void onStop() {
        super.onStop();
        note = collectNote();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        publisher.notifySingle(note);
    }

    private Note collectNote(){
        String title = this.title.getText().toString();
        String description = this.description.getText().toString();
        Date date = getDateFromDatePicker();
        boolean favorite;
        if (note != null){
            favorite = note.isFavorite();
        } else {
            favorite = false;
        }
        return new Note(title, date , description, favorite);
    }

    private Date getDateFromDatePicker() {
        Calendar cal = Calendar.getInstance();
        cal.set(Calendar.YEAR, this.datePicker.getYear());
        cal.set(Calendar.MONTH, this.datePicker.getMonth());
        cal.set(Calendar.DAY_OF_MONTH, this.datePicker.getDayOfMonth());
        return cal.getTime();
    }

    private void initView(View view) {
        title = view.findViewById(R.id.inputTitle);
        description = view.findViewById(R.id.inputDescription);
        datePicker = view.findViewById(R.id.inputDate);
    }

    private void populateView(){
        title.setText(note.getTitle());
        initDatePicker(note.getDate());
        description.setText(note.getDescription());

    }

    private void initDatePicker(Date date) {
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);
        this.datePicker.init(calendar.get(Calendar.YEAR),
                calendar.get(Calendar.MONTH),
                calendar.get(Calendar.DAY_OF_MONTH),
                null);
    }
}

File: app/src/main/java/com/tematikhonov/androidgeekbrainssecondapp/ui/NoteFragment.kt
package com.tematikhonov.androidgeekbrainssecondapp.ui

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.DatePicker
import androidx.fragment.app.Fragment
import com.google.android.material.textfield.TextInputEditText
import com.tematikhonov.androidgeekbrainssecondapp.MainActivity
import com.tematikhonov.androidgeekbrainssecondapp.R
import com.tematikhonov.androidgeekbrainssecondapp.data.Note
import com.tematikhonov.androidgeekbrainssecondapp.data.Publisher
import java.util.*

class NoteFragment : Fragment() {
    private var note: Note? = null
    private var publisher: Publisher? = null
    private var title: TextInputEditText? = null
    private var description: TextInputEditText? = null
    private var datePicker: DatePicker? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            note = arguments!!.getParcelable(ARG_CARD_DATA)
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        val activity = context as MainActivity
        publisher = activity.publisher
    }

    override fun onDetach() {
        publisher = null
        super.onDetach()
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_note, container, false)
        initView(view)
        if (note != null) {
            populateView()
        }
        return view
    }

    override fun onStop() {
        super.onStop()
        note = collectNote()
    }

    override fun onDestroy() {
        super.onDestroy()
        publisher!!.notifySingle(note)
    }

    private fun collectNote(): Note {
        val title = title!!.text.toString()
        val description = description!!.text.toString()
        val date = dateFromDatePicker
        val favorite: Boolean
        favorite = if (note != null) {
            note!!.isFavorite
        } else {
            false
        }
        return Note(title, date, description, favorite)
    }

    private val dateFromDatePicker: Date
        private get() {
            val cal = Calendar.getInstance()
            cal[Calendar.YEAR] = datePicker!!.year
            cal[Calendar.MONTH] = datePicker!!.month
            cal[Calendar.DAY_OF_MONTH] = datePicker!!.dayOfMonth
            return cal.time
        }

    private fun initView(view: View) {
        title = view.findViewById(R.id.inputTitle)
        description = view.findViewById(R.id.inputDescription)
        datePicker = view.findViewById(R.id.inputDate)
    }

    private fun populateView() {
        title!!.setText(note!!.title)
        initDatePicker(note!!.date)
        description!!.setText(note!!.description)
    }

    private fun initDatePicker(date: Date) {
        val calendar = Calendar.getInstance()
        calendar.time = date
        datePicker!!.init(calendar[Calendar.YEAR],
                calendar[Calendar.MONTH],
                calendar[Calendar.DAY_OF_MONTH],
                null)
    }

    companion object {
        private const val ARG_CARD_DATA = "Param_CardData"
        fun newInstance(note: Note?): NoteFragment {
            val fragment = NoteFragment()
            val args = Bundle()
            args.putParcelable(ARG_CARD_DATA, note)
            fragment.arguments = args
            return fragment
        }

        fun newInstance(): NoteFragment {
            return NoteFragment()
        }
    }
}